Основная логика проектов

mrnght opened the pull request.

mrnght2 committed:
добавил ProjectController

mrnght2 committed:
добавил ProjectService и реализовал в нем логику создания, обновления, получения и фильтрации проектов

mrnght2 committed:
BJS2-84035
покрыл тестами ProjectServiceImpl и ProjectController, добавил документацию

JekaCAP commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,58 @@
+        return ResponseEntity.ok().build();
+    }
+
+    @PutMapping("/{projectId}/update")
```
Как вариант, можно просто оставить {projectId} и в postman обращаться на PUT (http://localhost:8080/projects/{id})

JekaCAP commented on src/main/java/faang/school/projectservice/dto/client/project/CreateProjectDto.java:
```diff
@@ -0,0 +1,22 @@
+import java.math.BigInteger;
+import java.util.List;
+
+public record CreateProjectDto(
```
Record не защищает поля от null например. Стоит добавить валидацию полей

JekaCAP commented on src/main/java/faang/school/projectservice/dto/client/project/UpdateProjectDto.java:
```diff
@@ -0,0 +1,22 @@
+import java.math.BigInteger;
+import java.util.List;
+
+public record UpdateProjectDto(
```
Record не защищает поля от null например. Стоит добавить валидацию полей

JekaCAP commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+            mapper.update(projectDto, project.get());
+            repository.save(project.get());
+        } else if (!visibilityFilter(project.get())) {
+            throw new RuntimeException("У пользователя нет доступа к указанному проекту");
```
Исключения одного типа. Ошибка доступа НЕ равно ошибке доступа

JekaCAP commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+    @Override
+    @Transactional
+    public void updateProject(long id, UpdateProjectDto projectDto) {
+        Optional<Project> project = repository.findById(id);
```
Ты используешь Optional, чтобы защититься от null? При этом проверка у тебя всё равно ручная. 
Использовать метод isPresent, это антипаттерн. Лучше завернуть в     Project project = repository.findById(id)
        .orElseThrow(() -> new ProjectNotFoundException(id));

JekaCAP commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+    @Override
+    @Transactional
+    public void createProject(CreateProjectDto projectDto) {
+        Optional<Project> isHaveProjectWithSameName =
```
Это так же, рабочий вариант, что и метод ниже. Но ты не используешь преимущества Optional
ты не используешь ни map, ни orElseThrow, ни ifPresent. Теоретически можно метод  boolean existsByOwnerIdAndName(Long ownerId, String name); вынести в репозиторий проектов и вызывать его. Тогда код будет чище и короче.

RinatBeybutov replied:
Хорошая идея написать метод в репозитории

JekaCAP reviewed (commented):
Привет! Отличный код. Несколько небольших замечаний оставил.

mrnght2 committed:
BJS2-84035
немножко оптимизировал использование optional

mrnght2 committed:
BJS2-84035
добавил кастомное исключение

mrnght2 committed:
BJS2-84035
немножко исправил тест обновления проекта

RinatBeybutov commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,59 @@
+public class ProjectController {
+    private final ProjectService service;
+
+    @PostMapping("/create")
```
/create - лишнее. Метод Post говорит о том, что мы создаем в этом методе.
Общепринято, что метод создания возвращает созданный объект

RinatBeybutov commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @PutMapping("/{projectId}")
+    public ResponseEntity<Void> updateProject(@PathVariable long projectId,
```
обновление тоже возвращает обновленный объект

RinatBeybutov commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,59 @@
+        return ResponseEntity.ok(service.getProjectsFilteredByName());
+    }
+
+    @GetMapping()
```
пустые скобки убрать

RinatBeybutov commented on src/main/java/faang/school/projectservice/dto/client/project/ProjectDto.java:
```diff
@@ -0,0 +1,14 @@
+import faang.school.projectservice.model.ProjectStatus;
+import faang.school.projectservice.model.ProjectVisibility;
+
+public record ProjectDto(
```
ProjectDto -> ProjectViewDto

RinatBeybutov commented on src/main/java/faang/school/projectservice/dto/client/project/CreateProjectDto.java:
```diff
@@ -0,0 +1,33 @@
+import java.math.BigInteger;
+import java.util.List;
+
+public record CreateProjectDto(
```
CreateProjectDto -> ProjectCreateDto

RinatBeybutov commented on src/main/java/faang/school/projectservice/dto/client/project/UpdateProjectDto.java:
```diff
@@ -0,0 +1,33 @@
+import java.math.BigInteger;
+import java.util.List;
+
+public record UpdateProjectDto(
```
UpdateProjectDto -> ProjectUpdateDto

RinatBeybutov commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+public interface ProjectMapper {
+
+    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
+    Project toProject(CreateProjectDto projectDto);
```
toProject -> toEntity

RinatBeybutov commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,22 @@
+
+    void update(UpdateProjectDto projectDto, @MappingTarget Project entity);
+
+    ProjectDto toProjectDto(Project project);
```
toProjectDto -> toViewDto

RinatBeybutov commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    @Transactional
+    public void createProject(CreateProjectDto projectDto) {
+        Optional<Project> isHaveProjectWithSameName =
```
isHaveProjectWithSameName - это название хорошо подходит для boolean переменной, а не для Optional<project>.

Поэтому лучше назвать optionalProject

RinatBeybutov commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    @Transactional
+    public void updateProject(long id, UpdateProjectDto projectDto) {
+        Project project = repository.findById(id).orElseThrow(() ->
```
посмотри методы репозитория, может там уже есть метод, который кидает ошибку

RinatBeybutov commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,129 @@
+    @Transactional
+    public List<ProjectDto> getProjectsFilteredByStatus(ProjectDto projectDto) {
+        List<Project> projectList = repository.findAll();
+        log.info("Получение списка проектов, отсортированных по статусу {}",
```
сортировка - это расположение в определенном порядке
фильтрация - выдаче не всех записей, а только тех, которые удовлетворяют какому-то условию

У тебя путаница с понятиями

RinatBeybutov commented on src/main/java/faang/school/projectservice/service/ProjectServiceImpl.java:
```diff
@@ -0,0 +1,129 @@
+
+    @Override
+    @Transactional
+    public List<ProjectDto> getProjectsFilteredByName() {
```
нужно совместить в один метод получение всех записей, фильтрацию по статусу и по имени. Продумай архитектуру методов

mrnght2 committed:
BJS2-84035
косметические правки

mrnght2 committed:
BJS2-84035
создал систему фильтрации и переписал методы сервиса с ее использованием, добавил утильный класс
